editor/src/components/editor/store/store-hook.ts

```diff
@@ -224,3 +315,159 @@ export function useSelectorWithCallback<U>(
+      b.editor,
+    )
+  },
+  selectedViews: (a: SelectedViewsSubstate, b: SelectedViewsSubstate) =>
+    keysEquality(['selectedViews'], a.editor, b.editor),
+  focusedElement: (a: FocusedElementPathSubstate, b: FocusedElementPathSubstate) =>
+    keysEquality(['focusedElementPath'], a.editor, b.editor),
+  highlightedHoveredViews: (a: HighlightedViewsSubstate, b: HighlightedViewsSubstate): boolean => {
+    return (
+      // a.editor.selectedViews === b.editor.selectedViews &&
+      a.editor.highlightedViews === b.editor.highlightedViews &&
+      a.editor.hoveredViews === b.editor.hoveredViews
+    )
+  },
+  projectContents: (a: ProjectContentSubstate, b: ProjectContentSubstate) => {
+    return a.editor.projectContents === b.editor.projectContents
+  },
+  canvas: (a: CanvasSubstate, b: CanvasSubstate): boolean => {
+    return keysEquality(canvasSubstateKeys, a.editor.canvas, b.editor.canvas)
+  },
+  canvasOffset: (a: CanvasOffsetSubstate, b: CanvasOffsetSubstate) => {
+    return keysEquality(
+      ['realCanvasOffset', 'roundedCanvasOffset', 'scale'],
+      a.editor.canvas,
+      b.editor.canvas,
+    )
+  },
+  derived: (a: { derived: DerivedState }, b: { derived: DerivedState }) => {
+    return a.derived === b.derived
+  },
+  restOfEditor: (a: { editor: RestOfEditorState }, b: { editor: RestOfEditorState }) => {
+    return keysEquality(restOfEditorStateKeys, a.editor, b.editor)
+  },
+  restOfStore: (
+    a: Omit<EditorStorePatched, 'editor' | 'derived'>,
+    b: Omit<EditorStorePatched, 'editor' | 'derived'>,
+  ) => {
+    return keysEquality(restOfStoreKeys, a, b)
+  },
+  /** @deprecated hurts performance, please avoid using it */
+  fullStore: (a: EditorStorePatched, b: EditorStorePatched) => {
+    return a === b
+  },
+  theme: (a: ThemeSubstate, b: ThemeSubstate) =>
+    a.userState.themeConfig === b.userState.themeConfig,
+  github: (a: GithubSubstate, b: GithubSubstate) => {
+    return keysEquality(githubSubstateKeys, a.editor, b.editor)
+  },
+  builtInDependencies: (a: BuiltInDependenciesSubstate, b: BuiltInDependenciesSubstate) => {
+    return a.builtInDependencies === b.builtInDependencies
+  },
+  userState: (a: UserStateSubstate, b: UserStateSubstate) => {
+    return a.userState === b.userState
+  },
+  canvasAndMetadata: (a: CanvasAndMetadataSubstate, b: CanvasAndMetadataSubstate) => {
+    return (
+      a.editor.jsxMetadata === b.editor.jsxMetadata &&
+      keysEquality(canvasSubstateKeys, a.editor.canvas, b.editor.canvas)
+    )
+  },
+} as const
+
+export const SubstateEqualityFns: {
+  [key in StoreKey]: (a: Substates[key], b: Substates[key]) => boolean
+} = Substores
+
+export type UtopiaStores = { [key in StoreKey]: Store<Substates[key]> }
+export interface StoresAndSetState {
+  stores: UtopiaStores
+  setState: (store: EditorStorePatched) => void
+  getState: () => EditorStorePatched
+}
+
+export const createStoresAndState = (initialEditorStore: EditorStorePatched): StoresAndSetState => {
```

Comment:
I think this file, and this function in particular, could do with copious amounts of comments to explain what is happening and why here, as at this point it requires a fair amount of knowledge of how zustand works